File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/AuditLogViewModel.swift
import Foundation
import Observation
import FlagshipAPI

/// Backs the dedicated full-page audit-log viewer (P5). Mirrors the
/// canonical webapp `views/audit-log.js`: a paginated, time-sorted list
/// of every account-level audit event from flagshipserver.com
/// (`GET /api/users/:u/audit?since=&limit=`).
///
/// Pagination note: the `.com` endpoint returns rows DESC by `seq`,
/// `since` is an EXCLUSIVE LOWER bound, and `limit` is server-capped at
/// 50. We page by growing the request window (`limit += pageSize`) and
/// re-reading from `since: 0`, then merging by `seq`. "Load more" stops
/// once the server returns fewer rows than the requested window (no more
/// history) or the 50-row cap is reached. Events are always presented
/// newest-first.
@MainActor
@Observable
public final class AuditLogViewModel {
    public enum Status: Equatable {
        case idle
        case loading
        case loaded
        case failed(String)
    }

    /// Server-side per-request cap (mirrors MAX_LIMIT in
    /// packages/control-plane/src/auditEvents.ts).
    public static let maxLimit = 50

    public private(set) var status: Status = .idle
    /// Events newest-first (descending `seq`).
    public private(set) var events: [AuditEvent] = []
    /// Whether another "load more" page is available.
    public private(set) var canLoadMore: Bool = false
    public private(set) var loadingMore: Bool = false

    private let server: any FlagshipServerClient
    private let username: String
    private let pageSize: Int
    /// The window we last requested (`limit`). Grows on each load-more.
    private var requestedLimit: Int

    public init(
        server: any FlagshipServerClient,
        username: String,
        pageSize: Int = 30
    ) {
        self.server = server
        self.username = username
        self.pageSize = max(1, pageSize)
        self.requestedLimit = max(1, pageSize)
    }

    /// Fresh load (also the pull-to-refresh path). Resets the window.
    public func load() async {
        status = .loading
        requestedLimit = pageSize
        await fetch(limit: requestedLimit, replacing: true)
    }

    /// Grow the window by one page and re-read. No-op while a load is in
    /// flight, when there's nothing more to fetch, or once the server cap
    /// is hit.
    public func loadMore() async {
        guard canLoadMore, !loadingMore, status == .loaded else { return }
        loadingMore = true
        defer { loadingMore = false }
        requestedLimit = min(Self.maxLimit, requestedLimit + pageSize)
        await fetch(limit: requestedLimit, replacing: true)
    }

    private func fetch(limit: Int, replacing: Bool) async {
        guard !username.isEmpty else {
            events = []
            canLoadMore = false
            status = .loaded
            return
        }
        do {
            let resp = try await server.listAuditEvents(
                username: username, sinceSeq: 0, limit: limit
            )
            // Defensive: the Mock + Worker both return DESC, but sort
            // here so the screen never depends on transport ordering.
            let sorted = resp.events.sorted { $0.seq > $1.seq }
            events = sorted
            // More history is available iff the server filled the window
            // AND we haven't hit its hard cap.
            canLoadMore = sorted.count >= limit && limit < Self.maxLimit
            status = .loaded
        } catch {
            // Keep any already-shown events on a load-more failure so the
            // list doesn't blank out; only a fresh load surfaces .failed.
            if replacing && events.isEmpty {
                status = .failed(error.localizedDescription)
            } else {
                status = .loaded
            }
        }
    }

    // MARK: - Event kind → label / icon mapping (docs/revocation-ui.md)

    /// Human label for an event kind. Mirrors the inline Activity feed +
    /// docs/revocation-ui.md. Unknown kinds fall back to the raw string.
    public static func label(for kind: String) -> String {
        switch kind {
        case "device-disconnected": return "Disconnected device"
        case "device-replaced":     return "Replaced device"
        case "device-added":        return "Added device"
        case "wipe-restart":        return "Wiped & restarted account"
        case "recovery-set-up":     return "Set up recovery"
        case "recovery-rotated":    return "Rotated recovery passkey"
        case "app-renamed":         return "Renamed app URL"
        default:                    return kind
        }
    }

    /// SF Symbol for an event kind (per docs/revocation-ui.md).
    public static func icon(for kind: String) -> String {
        switch kind {
        case "device-disconnected": return "lock.open.trianglebadge.exclamationmark"
        case "device-replaced":     return "arrow.triangle.2.circlepath.circle"
        case "device-added":        return "plus.circle"
        case "wipe-restart":        return "trash.fill"
        case "recovery-set-up":     return "key.horizontal.fill"
        case "recovery-rotated":    return "arrow.triangle.2.circlepath"
        case "app-renamed":         return "link.circle"
        default:                    return "circle.fill"
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/AuditLogViewModel.kt
// P5 — dedicated audit-log viewer.
//
// Pages the account-level audit feed off .com
// (FlagshipServerClient.listAuditEvents → GET /api/users/:u/audit).
// Mirrors webapp views/audit-log.js (the full-page list). The Worker's
// `since` is an EXCLUSIVE LOWER bound that returns the newest `limit`
// rows (ORDER BY seq DESC, capped at 50), so "load more" grows the
// requested window rather than walking a cursor — the only paging the
// endpoint supports. The kind→label map mirrors docs/revocation-ui.md.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.AuditEvent
import com.flagshipserver.app.api.FlagshipServerClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Server-side cap on `limit` (packages/control-plane/src/auditEvents.ts
 *  MAX_LIMIT). Once the window reaches this, no further events can be
 *  fetched through this endpoint and the "load more" affordance hides. */
const val AUDIT_LOG_MAX_WINDOW = 50
const val AUDIT_LOG_PAGE_SIZE = 20

/** Human-readable label for each audit event kind. Pinned to
 *  docs/revocation-ui.md so the log reads consistently regardless of
 *  which subsystem authored the event. Mirrors the iOS + webapp maps. */
fun auditEventLabel(kind: String): String = when (kind) {
    "device-disconnected" -> "Disconnected device"
    "device-replaced"     -> "Replaced device"
    "device-added"        -> "Added device"
    "wipe-restart"        -> "Wiped & restarted account"
    "recovery-set-up"     -> "Set up recovery"
    "recovery-rotated"    -> "Rotated recovery passkey"
    "app-renamed"         -> "Renamed app URL"
    "server-created"      -> "Created server"
    "server-online"       -> "Server came online"
    else                  -> kind
}

data class AuditLogPage(
    val events: List<AuditEvent>,
    /** True while the server might hold more rows than we've fetched —
     *  i.e. the last fetch filled the requested window AND the window is
     *  still below the server's hard cap. */
    val canLoadMore: Boolean,
    /** True while a load-more fetch is in flight (the first-page fetch
     *  uses LoadingState.Loading instead). */
    val loadingMore: Boolean = false,
)

class AuditLogViewModel(
    private val client: FlagshipServerClient,
    private val username: () -> String?,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    private val _state = MutableStateFlow<LoadingState<AuditLogPage>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<AuditLogPage>> = _state.asStateFlow()

    private var window = AUDIT_LOG_PAGE_SIZE

    /** Initial load (or pull-to-refresh): reset the window + fetch the
     *  newest page. */
    fun load() = scope.launch {
        window = AUDIT_LOG_PAGE_SIZE
        _state.value = LoadingState.Loading
        fetch()
    }

    /** Grow the window by one page + re-fetch. No-op if a fetch is in
     *  flight or the window already hit the server cap. */
    fun loadMore() = scope.launch {
        val current = _state.value
        if (current !is LoadingState.Loaded || !current.value.canLoadMore || current.value.loadingMore) {
            return@launch
        }
        window = (window + AUDIT_LOG_PAGE_SIZE).coerceAtMost(AUDIT_LOG_MAX_WINDOW)
        _state.value = LoadingState.Loaded(current.value.copy(loadingMore = true))
        fetch()
    }

    private suspend fun fetch() {
        val u = username()
        if (u.isNullOrEmpty()) {
            _state.value = LoadingState.Loaded(AuditLogPage(events = emptyList(), canLoadMore = false))
            return
        }
        runCatching { client.listAuditEvents(u, sinceSeq = 0, limit = window).events }
            .fold(
                onSuccess = { events ->
                    val canLoadMore = events.size >= window && window < AUDIT_LOG_MAX_WINDOW
                    _state.value = LoadingState.Loaded(AuditLogPage(events = events, canLoadMore = canLoadMore))
                },
                onFailure = { t ->
                    _state.value = LoadingState.Failed(t.message ?: "couldn't load the audit log")
                },
            )
    }
}
